### PR TITLE
keymap: Support <D-...> (super/command key).

### LINKED ISSUE #2190: Support for cmd key <D-...>?
I tried searching if this was answered already, but the answers seemed outdated as the new ui system was pulled recently and I didn't see it mentioned in [the wiki](https://github.com/neovim/neovim/wiki/Differences-from-vim#features). Also, searching for "command key" is pretty useless. So has that support been enabled yet (or will it ever)?

### PR DESCRIPTION
**Pull request by justinmk**
Adds support for:
- api:`vim_input("<D-a>")`
- "`:nnoremap <C-D-S-...>`" and permutations thereof

UIs must capture the modifier and send it as "`<D-...>`" to vim_input().

Note: Before this commit, any arbitrary "`:nnoremap <{foo}-{bar}>`"
mapping could already be invoked with `feedkeys("\<{foo}-{bar}>")`. This
commit supports "D-" as a modifier that can be combined with "C-", "A-",
"S-" in any order.

For non-GUI (terminal) support, user must:

```
:set <D-a>={CSI sequence}
```

then send the {CSI sequence} from their terminal. But this does not work
yet (regression #2204).

Closes #2190

### CODE REVIEW

File: src/nvim/edit.c [hunk 1]
```diff
@@ -4967,6 +4962,9 @@ insert_special (
    * Only use mod_mask for special keys, to avoid things like <S-Space>,
    * unless 'allow_modmask' is TRUE.
    */
+  if (mod_mask & MOD_MASK_CMD) {  // Command-key never produces a normal key.
+    allow_modmask = true;
+  }
```
Comment (justinmk) [inline src/nvim/edit.c]: from Vim's gui_mac.c

File: src/nvim/keymap.c [hunk 2]
```diff
@@ -658,9 +654,11 @@ static int extract_modifiers(int key, int *modp)
 {
   int modifiers = *modp;
 
-  if ((modifiers & MOD_MASK_SHIFT) && ASCII_ISALPHA(key)) {
-    key = TOUPPER_ASC(key);
-    modifiers &= ~MOD_MASK_SHIFT;
+  if (!(modifiers & MOD_MASK_CMD)) {  // Command-key is special
```
Comment (justinmk) [inline src/nvim/keymap.c]: from Vim's gui_mac.c

**Comment by justinmk** [comment]
@equalsraf Looking at neovim-qt's `InputConv::modPrefix` I'm guessing this might work without any modifications? Is it also possible to capture the super ("Windows") key in gtk?

**Comment by equalsraf** [comment]
> Looking at neovim-qt's InputConv::modPrefix I'm guessing this might work without any modifications?

Yes for Mac OS X it should, cc @rkitover.

Gtk has GDK_SUPER_MASK, in [vim-gtk](https://github.com/vim/vim/blob/1357caf31f3e447d99baf8dd6a3243e2ab38567f/src/gui_gtk_x11.c#L898) is seems to map to META.

**Comment by justinmk** [comment]
@rogual Might be interested in this as well.

**Comment by rogual** [comment]
Thanks for the ping, will add this to nv.app

**Comment by rhysd** [comment]
@justinmk 

Thanks.  I will implement meta key to NyaoVim after next Neovim release.

**Comment by rhysd** [comment]
@justinmk 

This looks not to work in HEAD.  I set `:nnoremap <D-s> :echo "test"<CR>` as precondition.  Then I send `vim_input('<D-s>')` via RPC but nvim did not interpret it as Cmd+s.  In insert mode, `<D-s>` string is inserted as-is.

**Comment by justinmk** [comment]
@rhysd you need inoremap :)

**Comment by rhysd** [comment]
@justinmk Ah... sorry :flushed:

**Pull request merged**